### PR TITLE
feat: spancat

### PR DESCRIPTION
**Pull request by bdura**
## Description

This PR adds a SpanCat LLM Task. For the most part, this is a shameless copy of the NER taks.

### Types of change

This is a feature.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests (including the **external** tests), and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -290,6 +290,31 @@ labels = PERSON,ORGANISATION,LOCATION
...
+[components.llm.task]
+@llm_tasks = "spacy.SpanCat.v1"
+labels = PERSON,ORGANISATION,LOCATION
+examples = null
+```
+
+| Argument                  | Type                                    | Default      | Description                                                                                                                                  |
+| ------------------------- | --------------------------------------- | ------------ | -------------------------------------------------------------------------------------------------------------------------------------------- |
+| `labels`                  | `str`                                   |              | Comma-separated list of labels.                                                                                                              |
+| `span_key`                | `str`                                   | `"sc"`       | Key of the `Doc.spans` dict to save the spans under.                                                                                         |
```
Comment (adrianeboyd) [inline README.md:308]: I'm not sure where the naming needs to modified, but this should be `spans_key` not `span_key`.

Comment (bdura) [reply README.md:308]: Ooh nice catch

File: spacy_llm/tasks/spancat.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        """
+        super(SpanCatTask, self).__init__(
+            labels=labels,
+            examples=examples,
+            normalizer=normalizer,
+            alignment_mode=alignment_mode,
+            case_sensitive_matching=case_sensitive_matching,
+            single_match=single_match,
+        )
+        self._span_key = spans_key
```
Comment (adrianeboyd) [inline spacy_llm/tasks/spancat.py:83]: `spans_key`

File: spacy_llm/tasks/spancat.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                    case_sensitive=self._case_sensitive_matching,
+                    single_match=self._single_match,
+                )
+                for start, end in offsets:
+                    span = doc.char_span(
+                        start, end, alignment_mode=self._alignment_mode, label=label
+                    )
+                    if span is not None:
+                        spans.append(span)
+            doc.spans[self._span_key] = spans
```
Comment (adrianeboyd) [inline spacy_llm/tasks/spancat.py:105]: I think everything except this final assignment line should be shared with the NER task?

**Comment by bdura** [comment]
I made the inheritance structure clearer (in my mind): instead of having SpanCat inherit from NER, they both inherit from an `AbstractSpanTask`.

I also moved the templates to a dedicated submodule. Not sure about this – I do enjoy the syntax highlighting capabilities it enables though.

**Comment by kabirkhan** [comment]
It's worth monitoring this change to the NER prompt to include label definitions: https://github.com/explosion/spacy-llm/pull/84. It'll also need to be integrated with spancat but that's not necessarily a blocker for this PR.

File: README.md [hunk 4]
```diff
@@ -290,6 +290,31 @@ labels = PERSON,ORGANISATION,LOCATION
...
+| `normalizer`              | `Optional[Callable[[str], str]]`        | `None`       | Function that normalizes the labels as returned by the LLM. If `None`, defaults to `spacy.LowercaseNormalizer.v1`.                           |
+| `alignment_mode`          | `str`                                   | `"contract"` | Alignment mode in case the LLM returns entities that do not align with token boundaries. Options are `"strict"`, `"contract"` or `"expand"`. |
+| `case_sensitive_matching` | `bool`                                  | `False`      | Whether to search without case sensitivity.                                                                                                  |
+| `single_match`            | `bool`                                  | `False`      | Whether to match an entity in the LLM's response only once (the first hit) or multiple times.                                                |
+
+Except the `span_key` parameter, the SpanCat reuses the configuration from the NER task.
```
Comment (adrianeboyd) [inline README.md:315]:
```suggestion
Except for the `spans_key` parameter, the SpanCat reuses the configuration from the NER task.
```

File: spacy_llm/tasks/util/span_task.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+import jinja2
+from spacy.tokens import Doc, Span
+
+from ...compat import Literal
+from ...registry import lowercase_normalizer
+from .examples import SpanExample
+from .parsing import find_substrings
+
+
+class AbstractSpanTask:
```
Comment (adrianeboyd) [inline spacy_llm/tasks/util/span_task.py:12]: `Base` seemed more appropriate to me?

```suggestion
class BaseSpanTask:
```

File: spacy_llm/tests/tasks/test_spancat.py [hunk 6]
```diff
@@ -0,0 +1,464 @@
+# mypy: ignore-errors
```
Comment (adrianeboyd) [inline spacy_llm/tests/tasks/test_spancat.py:1]: Did you mean to leave this long-term?

File: spacy_llm/tasks/util/span_task.py (hunk 5 above)
Comment (rmitsch) [reply spacy_llm/tasks/util/span_task.py:12]: Or just `SpanTask`?

File: spacy_llm/tests/tasks/test_spancat.py (hunk 6 above)
Comment (rmitsch) [reply spacy_llm/tests/tasks/test_spancat.py:1]: Ideally we don't ignore errors file-wide.

**Comment by rmitsch** [comment]
> Not sure about this – I do enjoy the syntax highlighting capabilities it enables though.

This is fine.

File: spacy_llm/tests/tasks/test_spancat.py (hunk 6 above)
Comment (bdura) [reply spacy_llm/tests/tasks/test_spancat.py:1]: I copy-pasted the test suite for the NER. I'll change that in both files.

File: spacy_llm/tasks/util/span_task.py (hunk 5 above)
Comment (bdura) [reply spacy_llm/tasks/util/span_task.py]: Thanks for the suggestion, it's committed.

**Pull request merged**